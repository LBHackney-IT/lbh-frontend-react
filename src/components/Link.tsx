import classNames from "classnames";
import PropTypes from "prop-types";
import React from "react";

import { ComponentRegister } from "../ComponentRegister";

import "lbh-frontend/lbh/core/_links.scss";

/**
 * The proptypes for {@link Link}.
 */
export interface LinkProps {
  id?: string;
  className?: string;
  href: string;
  children: React.ReactNode;
}

/**
 * A component for creating an anchor linking to other pages. You can customize
 * the component used internally for integration with routers using
 * {@link ComponentRegister.init}. By default, it uses an `<a>` tag.
 */
export const Link: React.FunctionComponent<LinkProps> = (
  props: LinkProps
): JSX.Element => {
  const { id, className, href, children } = props;

  return (
    <ComponentRegister.Link
      id={id}
      className={classNames("govuk-link", "lbh-link", className)}
      href={href}
    >
      {children}
    </ComponentRegister.Link>
  );
};

Link.propTypes = {
  id: PropTypes.string,
  className: PropTypes.string,
  href: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired
};
