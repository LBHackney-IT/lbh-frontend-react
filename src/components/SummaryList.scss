$govuk-font-family: "Open Sans";
$govuk-include-default-font-face: false;

// Inherits directly from `govuk-frontend`, rather than `lbh-frontend`, due to
// `lbh-frontend` having no summary list component at time of implementation.
@import "node_modules/govuk-frontend/govuk/components/summary-list/_summary-list.scss";
