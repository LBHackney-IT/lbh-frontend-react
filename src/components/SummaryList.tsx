import classNames from "classnames";
import PropTypes from "prop-types";
import React from "react";

// Inherits directly from `govuk-frontend`, rather than `lbh-frontend`, due to
// `lbh-frontend` having no summary list component at time of implementation.
import "govuk-frontend/govuk/components/summary-list/_summary-list.scss";

/**
 * A unit of information to display in a row of a {@link SummaryList}.
 */
export interface SummaryListRow {
  /**
   * The description or label of the piece of information summarised by this
   * row.
   */
  key: string;

  /**
   * The piece of information itself.
   */
  value: string;
}

/**
 * The proptypes for {@link SummaryList}.
 */
export interface SummaryListProps {
  id?: string;
  className?: string;

  /**
   * An array describing the items to display.
   */
  rows: SummaryListRow[];
}

/**
 * A component for summarising information.
 *
 * It is intended to be used to present pairs of linked information in a list.
 */
export const SummaryList: React.FunctionComponent<SummaryListProps> = (
  props: SummaryListProps
): JSX.Element => {
  const { id, className, rows } = props;

  return (
    <dl
      id={id}
      className={classNames("govuk-summary-list lbh-summary-list", className)}
    >
      {rows.map(({ key, value }, index) => (
        <div
          key={index}
          className="govuk-summary-list__row lbh-summary-list__row"
        >
          <dt className="govuk-summary-list__key lbh-summary-list__key">
            {key}
          </dt>
          <dd className="govuk-summary-list__value lbh-summary-list__value">
            {value}
          </dd>
        </div>
      ))}
    </dl>
  );
};

SummaryList.propTypes = {
  id: PropTypes.string,
  className: PropTypes.string,
  rows: PropTypes.arrayOf(
    PropTypes.exact({
      key: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired
    }).isRequired
  ).isRequired
};
