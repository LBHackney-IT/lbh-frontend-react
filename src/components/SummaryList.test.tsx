import React from "react";
import { create } from "react-test-renderer";

import { SummaryList } from "./SummaryList";
import { Link } from "./Link";

it("renders correctly with all props", () => {
  const link = (
    <Link
      id="testId"
      className="class1"
      href="/"
      target="_blank"
      aria-label="test link"
      data-test="test data"
    >
      Link
    </Link>
  );
  const oneLink = [link];
  const twoLinks = [link, link];
  const component = create(
    <SummaryList
      id="test"
      className="class1 class2"
      rows={[
        {
          key: "TestKey",
          value: "TestValue",
          actions: oneLink,
        },
        { key: "TestKey2", value: "TestValue2", actions: twoLinks },
      ]}
      aria-label="test summary list"
      data-test="test data"
    />
  );

  expect(component).toMatchInlineSnapshot(`
    <dl
      aria-label="test summary list"
      className="govuk-summary-list lbh-summary-list class1 class2"
      data-test="test data"
      id="test"
    >
      <div
        className="govuk-summary-list__row lbh-summary-list__row"
      >
        <dt
          className="govuk-summary-list__key lbh-summary-list__key"
        >
          TestKey
        </dt>
        <dd
          className="govuk-summary-list__value lbh-summary-list__value"
        >
          TestValue
        </dd>
        <dd
          className="govuk-summary-list__actions"
        >
          <a
            aria-label="test link"
            className="govuk-link lbh-link class1"
            data-test="test data"
            href="/"
            id="testId"
            target="_blank"
          >
            Link
          </a>
        </dd>
      </div>
      <div
        className="govuk-summary-list__row lbh-summary-list__row"
      >
        <dt
          className="govuk-summary-list__key lbh-summary-list__key"
        >
          TestKey2
        </dt>
        <dd
          className="govuk-summary-list__value lbh-summary-list__value"
        >
          TestValue2
        </dd>
        <dd
          className="govuk-summary-list__actions"
        >
          <ul
            className="govuk-summary-list__actions-list"
          >
            <li
              className="govuk-summary-list__actions-list-item"
            >
              <a
                aria-label="test link"
                className="govuk-link lbh-link class1"
                data-test="test data"
                href="/"
                id="testId"
                target="_blank"
              >
                Link
              </a>
            </li>
            <li
              className="govuk-summary-list__actions-list-item"
            >
              <a
                aria-label="test link"
                className="govuk-link lbh-link class1"
                data-test="test data"
                href="/"
                id="testId"
                target="_blank"
              >
                Link
              </a>
            </li>
          </ul>
        </dd>
      </div>
    </dl>
  `);
});

it("renders correctly without optional props", () => {
  const component = create(
    <SummaryList rows={[{ key: "TestKey", value: "TestValue" }]} />
  );

  expect(component).toMatchInlineSnapshot(`
    <dl
      className="govuk-summary-list lbh-summary-list"
    >
      <div
        className="govuk-summary-list__row lbh-summary-list__row"
      >
        <dt
          className="govuk-summary-list__key lbh-summary-list__key"
        >
          TestKey
        </dt>
        <dd
          className="govuk-summary-list__value lbh-summary-list__value"
        >
          TestValue
        </dd>
      </div>
    </dl>
  `);
});

it("renders correctly with an empty array of rows", () => {
  const component = create(<SummaryList rows={[]} />);

  expect(component).toMatchInlineSnapshot(`
    <dl
      className="govuk-summary-list lbh-summary-list"
    />
  `);
});
