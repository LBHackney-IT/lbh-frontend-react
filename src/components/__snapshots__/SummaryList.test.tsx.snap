// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with all props 1`] = `
<dl
  className="govuk-summary-list lbh-summary-list class1 class2"
  id="test"
>
  <div
    className="govuk-summary-list__row lbh-summary-list__row"
  >
    <dt
      className="govuk-summary-list__key lbh-summary-list__key"
    >
      TestKey
    </dt>
    <dd
      className="govuk-summary-list__value lbh-summary-list__value"
    >
      TestValue
    </dd>
  </div>
  <div
    className="govuk-summary-list__row lbh-summary-list__row"
  >
    <dt
      className="govuk-summary-list__key lbh-summary-list__key"
    >
      TestKey2
    </dt>
    <dd
      className="govuk-summary-list__value lbh-summary-list__value"
    >
      TestValue2
    </dd>
  </div>
</dl>
`;

exports[`renders correctly with an empty array of rows 1`] = `
<dl
  className="govuk-summary-list lbh-summary-list"
/>
`;

exports[`renders correctly without optional props 1`] = `
<dl
  className="govuk-summary-list lbh-summary-list"
>
  <div
    className="govuk-summary-list__row lbh-summary-list__row"
  >
    <dt
      className="govuk-summary-list__key lbh-summary-list__key"
    >
      TestKey
    </dt>
    <dd
      className="govuk-summary-list__value lbh-summary-list__value"
    >
      TestValue
    </dd>
  </div>
</dl>
`;
